application/controllers/ServicesController.php

```diff
@@ -199,6 +202,85 @@ public function searchEditorAction()
         $this->setTitle(t('Adjust Filter'));
     }
 
+    public function gridAction()
+    {
+        $db = $this->getDb();
+
+        if ($this->params->has('noscript_apply')) {
+            $this->redirectNow($this->getRequest()->getUrl()->without('noscript_apply'));
+        }
+
+        $this->setTitle(t('Service Grid'));
+        $this->setAutorefreshInterval(15);
+        $query = Service::on($db)->with([
+            'state',
+            'host',
+            'host.state'
+        ]);
+
+        $gridcols = [
+            'host.display_name',
+            'host.name',
+            'service.name',
+            'service.display_name',
+            'service.state.is_handled',
+            'service.state.output',
+            'service.state.soft_state'
+        ];
+
+        $filter = (bool) $this->params->shift('problems', false) ?
+            \Icinga\Data\Filter\Filter::where('service.state.is_problem', 1) : null;
+
+        $this->filter($query);
+        $query->getSelectBase();
+
+        $this->applyRestrictions($query);
+
+        $this->view->problemToggle = $problemToggle = new Form(['method' => 'GET']);
+        $problemToggle->setUidDisabled();
+        $problemToggle->setTokenDisabled();
+        $problemToggle->setAttrib('class', 'filter-toggle inline icinga-controls');
+        $problemToggle->addElement('checkbox', 'problems', [
+            'disableHidden' => true,
+            'autosubmit'    => true,
+            'value'         => $filter !== null,
+            'label'         => $this->translate('Problems Only'),
+            'decorators'    => ['ViewHelper', ['Label', ['placement' => 'APPEND']]]
+        ]);
+
+        $query = new SimpleQuery($query->columns($gridcols)->assembleSelect());
```

Comment:
The `PivotTable` below should be able to handle `ipl\Orm\Query`. We don't use `SimpleQuery` in Icinga DB Web anymore.